kata-deploy/README.md

```diff
@@ -0,0 +1,131 @@
+    + [Kata cleanup](#kata-cleanup-)
+
+
+[kata-deploy](kata-deploy) provides a Dockerfile which contains all of the binaries
+and artifacts required to run Kata Containers, as well as reference daemonsets which can be utilized to install Kata Containers on a running Kubernetes cluster.
+
+Note, installation through daemonsets only succesfully installs `kata-containers.io/kata-runtime` on
+a node if it uses either containerd or CRI-O CRI-shims.
+
+## Quick start:
+
+### Install Kata on a running Kubernetes cluster
+
+```
+kubectl apply -f kata-rbac.yaml
+kubectl apply -f kata-deploy.yaml
+```
+
+### Run a sample workload
+
+Untrusted workloads can node-select based on ```kata-containers.io/kata-runtime=true```, and are
+run through ```kata-containers.io/kata-runtime``` if they are marked with the appropriate CRIO or containerd
+annotation:
+```
+CRIO:           io.kubernetes.cri-o.TrustedSandbox: "false"
+containerd:     io.kubernetes.cri.untrusted-workload: "true"
+```
+
+The following is a sample workload for running untrusted on a kata-enabled node:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+   annotations:
+    io.kubernetes.cri-o.TrustedSandbox: "false"
+    io.kubernetes.cri.untrusted-workload: "true"
+  labels:
+    env: test
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+    imagePullPolicy: IfNotPresent
+  nodeSelector:
+    kata-containers.io/kata-runtime: "true"
+```    
+
+To run:
+```
+kubectl apply -f examples/nginx-untrusted.yaml
+```
+
+Now, you should see the pod start. You can verify that the pod is making use of
+```kata-containers.io/kata-runtime``` by comparing the container ID observed with the following:
+```
+/opt/kata/bin/kata-containers.io/kata-runtime list
+kubectl describe pod nginx-untrusted
+```
+
+The following removes the test pod:
+```
+kubectl delete -f examples/nginx-untrusted.yaml
+```
+
+### Remove Kata from the Kubernetes cluster
+
+```
+kubectl delete -f kata-deploy.yaml
+kubectl apply -f kata-cleanup.yaml
+kubectl delete -f kata-cleanup.yaml
+kubectl delete -f kata-rbac.yaml
+```
+
+## kata-deploy Details
+
+### Dockerfile
+
+The Dockerfile used to create the container image deployed in the DaemonSet is provided here.
+This image contains all the necessary artifacts for running Kata Containers.
+
+Host artifacts:
+* kata-containers.io/kata-runtime: pulled from Kata GitHub releases page
+* kata-proxy: pulled from Kata GitHub releases page
+* kata-shim: pulled from Kata GitHub releases page
+* qemu-system-x86_64: statically built and included in this repo, based on Kata's QEMU repo
+* qemu/* : supporting binaries required for qemu-system-x86_64
+
+Virtual Machine artifacts:
+* kata-containers.img: pulled from Kata github releases page
+* vmliuz.container: pulled from Kata github releases page
+
+### Daemonsets and RBAC:
+
+A few daemonsets are introduced for kata-deploy, as well as an RBAC to facilitate
+appyling labels to the nodes.
```

Comment:
Is this optional? I'm a little worried about extra daemonsets with the power to label nodes. The kubelet permissions are restricted on a per node basis so a node can only edit its own record, but I don't think that applies to other service accounts?

Reply:
alternately, could you just use kubelets local credential? Then it would already exist and be restricted in exactly the needed way.

Reply:
@kfox1111 can you please provide suggested reading material and any examples that use it this way? RBAC noob :)

Reply:
maybe just pass /etc/kubernetes/kubelet.conf into the container and use that as the kubeconf? I don't think you would need rbac at all then.

Reply:
seems like that file refers to `system:node` and that's a whole lot more permissions?

```
kubectl describe clusterrole system:node
Name:         system:node
Labels:       kubernetes.io/bootstrapping=rbac-defaults
Annotations:  rbac.authorization.kubernetes.io/autoupdate=true
PolicyRule:
  Resources                                       Non-Resource URLs  Resource Names  Verbs
  ---------                                       -----------------  --------------  -----
  configmaps                                      []                 []              [get]
  endpoints                                       []                 []              [get]
  events                                          []                 []              [create patch update]
  nodes                                           []                 []              [create get list watch delete patch update]
  nodes/status                                    []                 []              [patch update]
  persistentvolumeclaims                          []                 []              [get]
  persistentvolumes                               []                 []              [get]
  pods                                            []                 []              [get list watch create delete]
  pods/eviction                                   []                 []              [create]
  pods/status                                     []                 []              [update]
  secrets                                         []                 []              [get]
  services                                        []                 []              [get list watch]
  tokenreviews.authentication.k8s.io              []                 []              [create]
  localsubjectaccessreviews.authorization.k8s.io  []                 []              [create]
  subjectaccessreviews.authorization.k8s.io       []                 []              [create]
  certificatesigningrequests.certificates.k8s.io  []                 []              [create get list watch]
  volumeattachments.storage.k8s.io                []                 []              [get]
```
vs
```
kubectl describe clusterrole "node-labeler"
Name:         node-labeler
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"rbac.authorization.k8s.io/v1","kind":"ClusterRole","metadata":{"annotations":{},"name":"node-labeler","namespace":""},"rules":[{"apiGrou...
PolicyRule:
  Resources  Non-Resource URLs  Resource Names  Verbs
  ---------  -----------------  --------------  -----
  nodes      []                 []              [get patch]
```

Reply:
yes, but no. Think of it in terms of what happens when a node gets compromised.

Adversary breaks out of container. Adversary has access to all secrets available on the single node they are on. From there, what can they do?

The system:node role has a special admin controller on the api side that restricts its behaviour to only the node it belongs to. so, for example, you can't evict pods on node B from node A.

But, with the role you created, an advisory can label any node. This could, for example, be used to label some more privileged nodes with labels that would attract the vulnerable workload the adversary used to get onto the machine in the first place. This would allow them to spread to all other machines, rather then just compromising some edge machines.

Ideally, the service account spec would allow node like restrictions on service accounts like it can do for node accounts. I don't think there is any plans currently to tackle that though.

Reply:
Found some pointers - 
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction
https://kubernetes.io/docs/reference/access-authn-authz/node/

Makes sense. Yeah node like restrictions on SA would have been ideal. @egernst WDYT?

Reply:
BTW, this is something we want to solve generically in Kubernetes (see https://github.com/kubernetes/kubernetes/issues/62747, DSP -> apiserver authz). I have a draft proposal for node-scoped roles, but it needs to be cleaned up before publishing.

Reply:
cool. thanks for the info.